src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/RestrictedInternalsVisibleToAnalyzer.cs

```diff
@@ -0,0 +1,248 @@
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var restrictedInternalsVisibleToMap = GetRestrictedInternalsVisibleToMap(compilationContext.Compilation);
+            if (restrictedInternalsVisibleToMap.IsEmpty)
+            {
+                return;
+            }
+
+            var namespaceToIsBannedMap = new ConcurrentDictionary<INamespaceSymbol, /*isBanned*/bool>();
+
+            // Verify all explicit type name specifications in declarations and executable code.
+            compilationContext.RegisterSyntaxNodeAction(
+                context =>
+                {
+                    var name = (TNameSyntax)context.Node;
+                    if (!IsInTypeOnlyContext(name) ||
+                        name.Parent is TNameSyntax)
+                    {
+                        // Bail out if we are not in type only context or the parent is also a name
+                        // which will be analyzed separately.
+                        return;
+                    }
+
+                    var typeInfo = context.SemanticModel.GetTypeInfo(name, context.CancellationToken);
+                    VerifySymbol(typeInfo.Type as INamedTypeSymbol, name, context.ReportDiagnostic);
+                },
+                NameSyntaxKinds);
+
+            // Verify all member usages in executable code.
+            compilationContext.RegisterOperationAction(
+                context =>
+                {
+                    ISymbol symbol;
+                    switch (context.Operation)
+                    {
+                        case IObjectCreationOperation objectCreation:
+                            symbol = objectCreation.Constructor;
+                            break;
+                        case IInvocationOperation invocation:
+                            symbol = invocation.TargetMethod;
+                            break;
+                        case IMemberReferenceOperation memberReference:
+                            symbol = memberReference.Member;
+                            break;
+                        default:
+                            throw new NotImplementedException($"Unhandled OperationKind: {context.Operation.Kind}");
+                    }
+
+                    VerifySymbol(symbol, context.Operation.Syntax, context.ReportDiagnostic);
+                },
+                OperationKind.ObjectCreation,
+                OperationKind.Invocation,
+                OperationKind.EventReference,
+                OperationKind.FieldReference,
+                OperationKind.MethodReference,
+                OperationKind.PropertyReference);
+
+            return;
+
+            // Local functions.
```

Comment:
>// Local functions. [](start = 12, length = 19)

I think it would be better to make these regular methods. It's not clear if the non-static local functions take dependency on local variables of the containig method.

Reply:
Done.